### PR TITLE
Migrate to vitest from jest

### PR DESCRIPTION
**Pull request by mattzcarey**
its newer, faster, works natively with the modern toolchain like typescript. 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist

- [x] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [x] I have added appropriate error handling
- [x] I have added or updated documentation as needed

### CODE REVIEW

**Comment by mattzcarey** [comment]
The main motivation to do this is that vitest works with ESM modules natively. This means we can use [MSW](https://mswjs.io/) to test MCP servers more effectively. MSW is a lovely tool and is pretty much standard practise when testing typescript servers.

**Comment by KKonstantinov** [comment]
Fully support this. Was discussed in https://github.com/modelcontextprotocol/typescript-sdk/pull/1015 and https://github.com/modelcontextprotocol/typescript-sdk/pull/916 but not progressed. Additionally other repos are already migrating (e.g. https://github.com/modelcontextprotocol/servers/pull/2884). It only makes sense that we do it here too!

**Comment by felixweinberger** [comment]
Support shipping this as this seems like a pareto improvement. Looks like there are some conflicts to resolve.

Looking at CI: 

CleanShot 2025-11-11 at 19 26 00

Not sure if these 2 are stalling because of this PR

File: src/server/streamableHttp.test.ts [hunk 1]
```diff
@@ -1369,16 +1376,19 @@ describe('StreamableHTTPServerTransport with resumability', () => {
         expect(storedEvent?.message).toMatchObject(notification);
     });
 
-    it('should store and replay MCP server tool notifications', async () => {
+    it('should store and replay MCP server tool notifications', { timeout: 10000 }, async () => {
```
Comment (felixweinberger) [inline src/server/streamableHttp.test.ts]: Why is the timeout needed here? Was it failing / flaking without this?
Doesn't look like the original had a timeout.

**Review by felixweinberger — APPROVED** [review]

File: src/server/streamableHttp.test.ts (hunk 1 above)
Comment (mattzcarey) [reply src/server/streamableHttp.test.ts]: it was a bit flaky. but I added a bunch more overhead with msw when i added this. Will check if I can remove.

Comment (mattzcarey) [reply src/server/streamableHttp.test.ts]: removed

File: src/client/auth.test.ts [hunk 2]
```diff
@@ -1110,17 +1119,10 @@ describe('OAuth Authorization', () => {
             });
 
             expect(tokens).toEqual(validTokens);
-            expect(mockFetch).toHaveBeenCalledWith(
```
Comment (pcarleton) [inline src/client/auth.test.ts:1113]: for my education, why was this necessary? does vitest not support this matcher pattern?

File: src/server/streamableHttp.test.ts [hunk 3]
```diff
@@ -1288,9 +1288,16 @@ describe('StreamableHTTPServerTransport with resumability', () => {
         ): Promise<StreamId> {
             const streamId = lastEventId.split('_')[0];
             // Extract stream ID from the event ID
-            // For test simplicity, just return all events with matching streamId that aren't the lastEventId
-            for (const [eventId, { message }] of storedEvents.entries()) {
-                if (eventId.startsWith(streamId) && eventId !== lastEventId) {
+            // Convert to array and find all events after the lastEventId
```
Comment (pcarleton) [inline src/server/streamableHttp.test.ts]: what happened here? is this related?

**Review by pcarleton — CHANGES_REQUESTED** [review]
overall looks good, ty for doing this. just had 2 questions about parts i didn't understand

File: src/client/auth.test.ts (hunk 2 above)
Comment (mattzcarey) [reply src/client/auth.test.ts:1113]: no it supports it fine my bad. reverting

File: src/server/streamableHttp.test.ts (hunk 3 above)
Comment (mattzcarey) [reply src/server/streamableHttp.test.ts]: its a bit more readable. functionally the same. reverting.

File: src/client/auth.test.ts (hunk 2 above)
Comment (mattzcarey) [reply src/client/auth.test.ts:1113]: hmmm seems vitest is alot stricter than jest in object matching. it sees an instance of Headers or URLSearchParams rather than whats inside when you do the top level objectContains...

Comment (mattzcarey) [reply src/client/auth.test.ts:1113]: basically you cant match the Headers object with one you made during the test because they are different objects

Comment (KKonstantinov) [reply src/client/auth.test.ts:1113]: Hey, this is true - you can't match it because it's a new instance of Headers (in vitest, this applies for any class instance).

But should you not be able to do:

```typescript
expect(mockFetch).toHaveBeenCalledWith(
                expect.objectContaining({
                    href: 'https://auth.example.com/token'
                }),
                expect.objectContaining({
                    method: 'POST',
                    headers: expect.objectContaining({
                        'Content-Type': 'application/x-www-form-urlencoded'
                    })
                })
            );
```

`expect(mockFetch.mock.calls[0][1])` would be flagged up by some linting rules (if ever added), because a [0] and a [1] element might not exist, so calling `.headers` on undefined is a possibility, where as toHaveBeenCalledWith is safe.

With vitest, we could achieve pretty much everything without having to resort to `.mock.calls` and similar - as these resort to testing actual implementation (e.g. in the test, we're hardcoding the exact order in which something gets called) as opposed to just testing the outcome.

**Comment by felixweinberger** [comment]
@mattzcarey pushed a fix for the failing CI
@pcarleton ready for a stamp

File: src/client/auth.test.ts (hunk 2 above)
Comment (pcarleton) [reply src/client/auth.test.ts:1113]: yea ideally we'd figure out how to do the matcher here, the [0][1] is a bit jank.  anyway, we can always fix later, let's move this on since it'll be a big merge conflict for other stuff if we leave it longer

**Review by pcarleton — APPROVED** [review]

**Comment by morozow** [comment]
I also noticed that Vitest is used across several repositories, and that there was a migration from Jest to Vitest. I left a more detailed comment in the server repository here: https://github.com/modelcontextprotocol/servers/pull/2884#issuecomment-4372281511
I recommend reconsidering this decision carefully, because there are potential stability and diagnostic concerns that should be reviewed before this pattern spreads further across repositories. The async/generator/mocked-function behavior mentioned there is only one small example of the broader concern. It should not be treated as the only reason. The main point is that Vitest can introduce different runtime and diagnostic behavior compared to Jest, and these differences may affect agents and developers who need stable, predictable test execution and clear failure signals.
For worldwide projects, including OSS projects, I would prefer to keep the test flow on Jest unless Vitest behavior is explicitly validated against these kinds of edge cases.

**Comment by morozow** [comment]
This migration introduced `vite` as a mandatory peer dep of vitest. The currently locked `vite@7.3.0` has two high-severity CVEs (GHSA-v2wj-q39q-566r, GHSA-p9ff-h696-f583), patched in ≥7.3.2. `pnpm.overrides` does not force the resolution for peer deps, so `pnpm audit --audit-level=high` fails on a clean checkout with no workaround short of adding `vite` as an explicit devDependency. Jest didn't have this problem — no bundler in the dep tree.
